jupyterhub/handlers/base.py

```diff
@@ -854,7 +854,12 @@ async def auth_to_user(self, authenticated, user=None):
             group_names = authenticated["groups"]
             if group_names is not None:
                 user.sync_groups(group_names)
-
+        # apply authenticator-managed roles
+        if self.authenticator.manage_roles:
+            print(authenticated)
+            auth_roles = authenticated.get("roles")
+            if auth_roles is not None:
+                user.sync_roles(auth_roles)
```

Comment:
Currently `sync_roles` (proposed in this PR) works differently from `user.sync_groups` and not as I would expect:
- both `user.sync_roles` and `user.sync_groups` will create groups/roles as needed
- `user.sync_groups` will add the user to the groups listed and remove from groups not listed
- `user.sync_roles` will never remove user from a role
- `user.sync_roles` will only add user to the roles listed
   - a) the role does not yet exist, AND
   - b) the role includes a list of users with the current user in it

This means that for a user `elephant` the authenticator response would need to look like:

```python
{
   'auth_state': ...,
   'name': 'elephant',
   'roles': [
      {
          'name': 'custom-role',
          'description': 'My custom role',
          'scopes': [...],
          'users': ['elephant']
      }
   ]
}
```

Further if we wanted other users to later get this role, all other users with that role would need to be listed in every response which I find rather inelegant and constraining.

I propose to rewrite `user.sync_roles` so that it adds the roles listed to the current user (whether they need creating or not) and _removes_ explicit user-role relation if the explicit roles defined for a user are no longer listed.